lib/active_merchant/billing/gateways/vpos.rb

```diff
@@ -146,15 +181,21 @@ def success_from(response)
       end
 
       def message_from(response)
-        response.dig('confirmation', 'extended_response_description') ||
-          response.dig('confirmation', 'response_description') ||
-          response.dig('confirmation', 'response_details') ||
-          response.dig('messages', 0, 'key')
+        %w(confirmation refund).each do |m|
+          message =
+            response.dig(m, 'extended_response_description') ||
+            response.dig(m, 'response_description') ||
+            response.dig(m, 'response_details')
+          return message if message
+        end
+        [response.dig('messages', 0, 'key'), response.dig('messages', 0, 'dsc')].join(':')
       end
 
       def authorization_from(response)
-        authorization_number = response.dig('confirmation', 'authorization_number')
-        shop_process_id = response.dig('confirmation', 'shop_process_id')
+        response_body = response.dig('confirmation') || response.dig('refund')
+        return unless response_body
+        authorization_number = response_body.dig('authorization_number') || response_body.dig('authorization_code')
```

Comment:
General curiosity, what is the difference between `authorization_number` and `authorization_code`? And is there ever a possibility where both could be returned in the response body?

Reply:
It seems like we get `authorization_number`s with `purchase`s (in a `confirmation` response) and `authorization_code`s with `credit`s (in a `refund` response). I am reasonably sure we'll never get both, since as far as I can tell we never get a response with both a `confirmation` and a `refund`.